A77-xds-rate-limiting-rlqs.md

```diff
@@ -0,0 +1,490 @@
+
+# Notes
+    Note right of xds: r1-4: routes <br />c1-2: unique filter configs
+%% LDS 1
+    xds ->> filter: LDS1<br />RLQS{r1=c1, r2=c2, r3=c2}
+    filter ->> cache: r1: getOrCreate(c1)
+    cache ->>+ e1: new RlqsFilterState(c1)
+    filter ->> cache: r2: getOrCreate(c2)
+    cache ->>+ e2: new RlqsFilterState(c2)
+    filter ->> cache: r3: getOrCreate(c2)
+    Note over filter: r1: RlqsFilterState(c1)<br/>r2: RlqsFilterState(c2)<br/>r3: RlqsFilterState(c2)
+%% RDS 1
+    xds ->> filter: RDS1<br />RLQS{r1=c2}
+    filter ->> cache: r1: getOrCreate(c2)
+    filter ->> cache: shutdownFilterState(c1)
+    cache -x e1: RlqsFilterState(c1).shutdown()
+    deactivate e1
+    Note over filter: r1: RlqsFilterState(c2)<br/>r2: RlqsFilterState(c2)<br/>r3: RlqsFilterState(c2)
+%% LDS 2
+    xds ->> filter: LDS2<br />RLQS{r3=c2, r4=c2}
+    filter ->> cache: r4: getOrCreate(c2)
+    Note over filter: r3: RlqsFilterState(c2)<br/>r4: RlqsFilterState(c2)
+%% End
+    deactivate e2
+```
+
+**LDS 1**
+
+In this example, the RLQS filter is configured for three routes: `r1`, `r2`,
+and `r3`. Each unique config generates a unique RLQS Filter
+State: `RlqsFilterState(c1)` for the config `c1`, and `RlqsFilterState(c2)` for
+the config `c2`. After processing the first LDS update, we've generated
+onCallHandlers for three routes:
+
+1. `r1`, referencing `RlqsFilterState(c1)`.
+2. `r2`, referencing `RlqsFilterState(c2)`.
+3. `r3`, also referencing `RlqsFilterState(c2)`.
+
+**RDS 1**
+
+RDS 1 updates RLQS config for the route `r1` so it's identical to config `c2`.
+We retrieve `RlqsFilterState(c2)` from the RLQS Cache and generate new
+onCallHandlers for route `r2`. `RlqsFilterState(c1)` is no longer referenced by
+any onCallHandler, and can be destroyed with all associated resources.
+
+**LDS 2**
+
+LDS 2 update removes `r1` and `r2`, and adds new route r4 with the config
+identical to `c2`. While onCallHandlers for routes `r1` and `r2` are
+destroyed, `RlqsFilterState(c2)` is still used by two onCallHandlers, so it's
+preserved in RLQS Cache.
+
+##### Future considerations
+
+With this proposal, the filter state is lost if change is made to the filter
+config, including updates to inconsequential fields such as deny response
+status. Additional logic can be introduced to handle updates to such fields
+while preserving the filter state.
+
+### Multithreading
+
+There are several mutex-synchronized operations executed in
+latency-sensitive `onCallHandler`:
+
+1. Inserting/reading a bucket from the bucket cache using `bucket_id`. Note
+   that `bucket_id` is represented as a `Map<String, String>`, which may
+   introduce complexities in efficient cache sharding for certain programming
+   languages.
+2. Incrementing `num_request_allowed`/`num_request_denied` bucket counters.
+
+Each gRPC implementation needs to consider what synchronization primitives are
+available in their language to minimize the thread lock time.
+
+### Code Samples
```

Comment:
These code snippets should be part of the descriptions of the RLQS filter (interceptor) and RLQS Filter State objects, respectively.